pulser/simulation/simulation.py

```diff
@@ -325,55 +463,125 @@ def get_hamiltonian(self, time: float) -> qutip.Qobj:
+            else self.config.solver_options
+
+        def _assign_meas_basis() -> str:
+            if hasattr(self._seq, '_measurement'):
+                return cast(str, self._seq._measurement)
             else:
-                if initial_state.shape != (self.dim**self._size,):
-                    raise ValueError("Incompatible shape of initial_state")
-                self._initial_state = qutip.Qobj(initial_state)
-        else:
-            # by default, initial state is "ground" state of g-r basis.
-            all_ground = [self.basis['g'] for _ in range(self._size)]
-            self._initial_state = qutip.tensor(all_ground)
-
-        result = qutip.sesolve(self._hamiltonian,
-                               self._initial_state,
-                               self.eval_times,
-                               progress_bar=progress_bar,
-                               options=qutip.Options(max_step=5,
-                                                     **options)
-                               )
-        meas_basis: Optional[str]
-        if hasattr(self._seq, '_measurement'):
-            meas_basis = cast(str, self._seq._measurement)
+                if self.basis_name in {'digital', 'all'}:
+                    return 'digital'
+                else:
+                    return 'ground-rydberg'
+
+        def _run_solver(as_subroutine: bool = False,
+                        measurement_basis: str = '') -> CleanResults:
+            """Returns CleanResults: Object containing evolution results."""
+            if not as_subroutine:
+                # CLEAN SIMULATION:
+                measurement_basis = _assign_meas_basis()
+            if self._collapse_ops:
+                # temporary workaround due to a qutip bug when using mesolve
+                liouvillian = qutip.liouvillian(self._hamiltonian,
+                                                self._collapse_ops)
+                result = qutip.mesolve(liouvillian,
+                                       self.initial_state,
+                                       self._eval_times_array,
+                                       progress_bar=progress_bar,
+                                       options=solv_ops)
+            else:
+                result = qutip.sesolve(self._hamiltonian,
+                                       self.initial_state,
+                                       self._eval_times_array,
+                                       progress_bar=progress_bar,
+                                       options=solv_ops)
+            return CleanResults(result.states, self._size,
+                                self.basis_name, self._eval_times_array,
+                                measurement_basis)
+
+        if self.config.noise:
+            # NOISY SIMULATION:
+            meas_basis = _assign_meas_basis()
+            if self.basis_name == 'all':
+                basis_name = 'digital'
+            else:
+                basis_name = self.basis_name
```

Comment:
I guess you can do this when you're giving it to `NoisyResults`, the placement here is a bit confusing